modules/core/src/smithy4s/schema/CachedSchemaCompiler.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.schema
+
+trait CachedSchemaCompiler[F[_]] {
+
+  type Cache
+  def createCache(): Cache
+  def fromSchema[A](schema: Schema[A]): F[A]
+  def fromSchema[A](schema: Schema[A], cache: Cache): F[A]
+
+}
+
+object CachedSchemaCompiler {
+
+  private[smithy4s] abstract class Impl[F[_]] extends CachedSchemaCompiler[F] {
+    protected type Aux[_]
+    type Cache = CompilationCache[Aux]
+
+    override final def fromSchema[A](schema: Schema[A]): F[A] =
+      fromSchema(schema, CompilationCache.nop[Aux])
+
+    def createCache(): Cache = CompilationCache.make[Aux]
+
+    private val globalCache: Cache = createCache()
+    implicit def derivedImplicitInstance[A](implicit
```

Comment:
Here we're trusting that it's okay enough to use a global cache, which is consistent in the previous usage. This is because when using implicits, the schemas are somewhat guaranteed to be tied to types, therefore they are "uniquely" instantiated for that usecase, whereas in our dynamic interpreters, we often apply a hint-transformation pre-processing step, which implies non-unique instances of schemas, which is the reason for the memory leak.